apps/Test App/Test App/Fake Server Backend/FakeBackend.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  FakeBackend.swift
+//  Test App
+//
+//  Created by Emily Dixon on 5/8/23.
+//
+
+import Foundation
+
+/// This class "fakes" the server backend necessary to compvare an upload workflow.
+/// In your production use case, a backend server should take care of creating upload URLs
+///
+/// **You should never build Mux server API ceredentials into a real app**. We do it in this example for brevity only
+class FakeBackend {
+    
+    func createDirectUpload() async throws -> URL {
+        let request = try {
+            var req = try URLRequest(url:fullURL(forEndpoint: "uploads"))
+            req.httpBody = try jsonEncoder.encode(CreateUploadPost())
+            req.httpMethod = "POST"
+            req.addValue("application/json", forHTTPHeaderField: "Content-Type")
+            req.addValue("application/json", forHTTPHeaderField: "accept")
+            
+            let basicAuthCredential = "\(MUX_ACCESS_TOKEN_ID):\(MUX_ACCESS_SECRET_KEY)".data(using: .utf8)!.base64EncodedString()
+            req.addValue("Basic \(basicAuthCredential)", forHTTPHeaderField: "Authorization")
+            
+            return req
+        }()
+        
+        let (data, response) = try await urlSession.data(for: request)
+        let httpResponse = response as! HTTPURLResponse
+        if (200...299).contains(httpResponse.statusCode) {
+            let responseData = try jsonDecoder.decode(CreateUploadResponseContainer.self, from: data).data
+            guard let uploadURL = URL(string:responseData.url) else {
+                throw CreateUploadError(message: "invalid upload url")
+            }
+            return uploadURL
+        } else {
+            self.logger.error("Upload POST failed: HTTP \(httpResponse.statusCode):\n\(String(decoding: data, as: UTF8.self))")
+            throw CreateUploadError(message: "Upload POST failed: HTTP \(httpResponse.statusCode):\n\(String(decoding: data, as: UTF8.self))")
+        }
+    }
+    
+    /// Generates a full URL for a given endpoint in the Mux Video public API
+    private func fullURL(forEndpoint: String) throws -> URL {
+        guard let url = URL(string: "https://api.mux.com/video/v1/\(forEndpoint)") else {
+            throw CreateUploadError(message: "bad endpoint")
+        }
+        return url
+    }
+    
+    private let logger = Test_AppApp.logger
+    
+    let urlSession: URLSession
+    let jsonEncoder: JSONEncoder
+    let jsonDecoder: JSONDecoder
+    
+    let MUX_ACCESS_TOKEN_ID = "2d8d1173-8489-4281-b530-3f17ac534f7e"
```

Comment:
Should these be environment variables?

Reply:
Prior art, both inside and out, is just to paste them in. I have worked with enough devs without console knowledge that I wouldn't want to require it

Reply:
These shouldn't be checked into a public location because they're API credentials and might end up misused.

But we should make the need to update these values more obvious so what about this? (Will result in a compilation failure with a custom error message directing to our dashboard.)

```
let MUX_ACCESS_TOKEN_ID = "INSERT_YOUR_MUX_ACCESS_TOKEN_HERE"
let MUX_ACCESS_SECRET_KEY = "INSERT_YOUR_MUX_SECRET_KEY_HERE"

#error (Create an Access Token and Secret in the Mux Dashboard and insert here. Then delete this line.”)
```

Reply:
I think ^ is less of a potential foot gun than environment variables

Reply:
I intended to replace these credentials with dummies before merging, I just needed them for testing.

I like the idea of putting an `#error` there tho, I can add that

Reply:
Er wait, Maybe that's not such a good idea. I'd like to be able to verify that the example app builds via CI at some point, so maybe we should just do the placeholders? I've already made that update

Reply:
> Er wait, Maybe that's not such a good idea. I'd like to be able to verify that the example app builds via CI at some point, so maybe we should just do the placeholders? I've already made that update

There's an internal solution we can reuse for creating CI builds, something like `"YOUR ACCESS TOKEN ID HERE"`, etc actually works better for that as well!